src/com/xilinx/rapidwright/router/VersalClockRouting.java

```diff
@@ -0,0 +1,613 @@
+                    RouteNode curr = q.poll();
+                    Node currNode = Node.getNode(curr);
+                    if (targetNode.equals(currNode)) {
+                        boolean inuse = false;
+                        for (PIP pip : curr.getPIPsBackToSourceByNodes()) {
+                            if (inuse) {
+                                assert(getNodeStatus.apply(pip.getStartNode()) == NodeStatus.INUSE);
+                                continue;
+                            }
+                            currPIPs.add(pip);
+                            NodeStatus status = getNodeStatus.apply(pip.getStartNode());
+                            if (status == NodeStatus.INUSE) {
+                                inuse = true;
+                                continue;
+                            }
+                            assert(status == NodeStatus.AVAILABLE);
+                        }
+                        sink.setRouted(true);
+                        visited.clear();
+                        continue nextPin;
+                    }
+
+                    for (Node downhill : currNode.getAllDownhillNodes()) {
+                        if (!allowedIntentCodes.contains(downhill.getIntentCode())) continue;
+                        if (!visited.add(downhill)) continue;
+                        if (used.contains(downhill)) continue;
+                        // have to allow those routethru-s NODE_IRI -> *
+                        if (routeThruHelper.isRouteThru(currNode, downhill) && downhill.getIntentCode() != IntentCode.NODE_IRI) continue;
+                        if (isNodeUnavailable.test(downhill)) continue;
+                        q.add(new RouteNode(downhill.getTile(), downhill.getWireIndex(), curr, curr.getLevel()+1));
+                    }
+                }
+                throw new RuntimeException("ERROR: Couldn't route LCB " + e.getKey() + " to Pin " + sink);
+            }
+
+            List<PIP> clkPIPs = clk.getPIPs();
+            for (PIP p : currPIPs) {
+                used.add(p.getStartNode());
+                used.add(p.getEndNode());
+                clkPIPs.add(p);
+            }
+        }
+    }
+
+    /**
+     * Routes from a GLOBAL_VERTICAL_ROUTE to horizontal distribution lines.
+     * @param clk The clock net to be routed.
+     * @param vroute The node to start the route.
+     * @param clockRegions Target clock regions.
+     * @param down To indicate if it is routing to the group of top clock regions.
+     * @return A list of RouteNodes indicating the reached horizontal distribution lines.
+     */
+    public static Map<ClockRegion, RouteNode> routeToHorizontalDistributionLines(Net clk,
+                                                                     RouteNode vroute,
+                                                                     Collection<ClockRegion> clockRegions,
+                                                                     boolean down,
```

Comment:
This `down` argument is never used?

Reply:
Yes. It's a parameter in UltraScale clock routing method.
In UltraScale symmetric clock routing, we use `vrouteUp` and `vrouteDown` to go up and down, and `down` is to distinguish these two nodes and compute the offset of Y-coordinate of `currCR` when we are using `vrouteDown`. 

However, I found that there is only one VROUTE node in vivado solutions for Versal designs, so there is no need to use `down` here.

Reply:
In Versal clock routing, we use `VDISTR_LVL{1, 2}` to go up and down.

Reply:
> However, I found that there is only one VROUTE node in vivado solutions for Versal designs, so there is no need to use down here.

Can you elaborate on what you mean by only one `VROUTE` node?

> In Versal clock routing, we use VDISTR_LVL{1, 2} nodes to go up and down.

Have you found (on the `xcvp1002`) that you only ever need to use `VDISTR` to go up?

Reply:
> Can you elaborate on what you mean by only one VROUTE node?

https://github.com/Xilinx/RapidWright/blob/c43c9fe7e5882d1d7b49a5fb1d477cd50799af2a/src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java#L281

Here we only have one VROUTE node for all sink pins. However, on US/US+ devices we have https://github.com/Xilinx/RapidWright/blob/c43c9fe7e5882d1d7b49a5fb1d477cd50799af2a/src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java#L208-L224 here we use two VROUTE nodes `vrouteUp` and `vrouteDown`.

Reply:
> Have you found (on the xcvp1002) that you only ever need to use VDISTR to go up?

BUFG_FABRIC sources are in the middle of the xcvp1002 so we may still need VDISTR to go down.